lib/broadway/acknowledger.ex

```diff
@@ -80,4 +82,45 @@ defmodule Broadway.Acknowledger do
     failed = Process.delete({ack_info, :failed}) || []
     acknowledger.ack(ack_ref, Enum.reverse(successful), Enum.reverse(failed))
   end
+
+  # Used by the processor and the batcher to maybe call c:handle_failed/2
+  # on failed messages.
+  @doc false
+  def maybe_handle_failed_messages(messages, module, context, size) do
+    if function_exported?(module, :handle_failed, 2) and messages != [] do
+      handle_failed_messages(messages, module, context, size)
+    else
+      messages
+    end
+  end
+
+  defp handle_failed_messages(messages, module, context, size) do
+    module.handle_failed(messages, context)
+  catch
+    kind, reason ->
+      Logger.error(Exception.format(kind, reason, System.stacktrace()))
+      failed = "due to an unhandled #{kind} in handle_failed/2"
+      Enum.map(messages, &Message.failed(&1, failed))
```

Comment:
They have already failed... should we mask the actual error or just return them as is?